lib/rspec/support/object_inspector.rb

```diff
@@ -3,23 +3,51 @@ module Support
     # Provide additional output details beyond what `inspect` provides when
     # printing Time, DateTime, or BigDecimal
     module ObjectInspector
+      DEFAULT_OPTIONS = {
+        :nested => false,
+        :joiner => ',',
+        :braces => true,
+        :trailing_comma => false
+      }
       # @api private
-      def self.inspect(object)
-        if Time === object
+      # rubocop:disable CyclomaticComplexity
+      # rubocop:disable MethodLength
+      def self.inspect(object, options=DEFAULT_OPTIONS)
+        case object
+        when Time
           format_time(object)
-        elsif defined?(DateTime) && DateTime === object
-          format_date_time(object)
-        elsif defined?(BigDecimal) && BigDecimal === object
-          "#{object.to_s 'F'} (#{object.inspect})"
-        elsif RSpec::Support.is_a_matcher?(object) && object.respond_to?(:description)
-          object.description
+        when Hash
+          format_hash(object, options)
```

Comment:
@myronmarston I agree with the options being messy, I was just trying to adapt the class to handle the cases that you want.
I think having slightly different functionality for an inspector for the general output and the differ output is fine, so long as we can reuse similar patterns between the both, possibly share a class they both inherit from.

I'm a little confused here, as I don't know where the description method originates, so I don't know what it does in terms of output - https://github.com/rspec/rspec-expectations/blob/v3.2.1/lib/rspec/matchers/composable.rb#L174
I don't understand how it handles arrays in particular.

> we want to replace how time objects are inspected when they are nested in arrays/hashes, but not worry about the formatting of the array or hash itself. It's nice to let ruby take care of that simply by using Array#inspect or Hash#inspect with the items in the data structures suitably replaced.

I'm not sure this will work for use in the base of:
```
[
  { key => TimeObject }
]
```
It seems that using `surface_descriptions_in` will see it's an array object, and then do whatever the description method does to display it. Perhaps it'll be more clear to me once I know what `description` does.

To be perfectly honest, part of my reasoning for removing Arrays from this branch was de-scoping. I just wanted to have to stop monkey patching `BigDecimal` when working inside of some tests :P, and I didn't expect it this pull request to last as long as it has. I want to see this through to completion, I just find myself with less and less time to work on it at the moment (and my desire for the original feature also waning). 

---

The issue I'm having is I want a recursive solution so that it inspects the objects inside objects inside objects, however I think the issue I'm having is that I was trying to do it all inside one object, and therefore handle the formatting.

Having a look at the links you provided, I think this might actually solve the problem I'm having, as it ensures that `.inspect` is only called **once** per array or hash.

``` ruby
class ObjectInspector
  def self.inspect(item)
    case
    when Array === item
      ArrayInspector.inspect(item).inspect
    when Hash === item
      Hash[ArrayInspector.inspect(item.to_a)].inspect
    when #other options
    else
      #default case
    end
  end
end

class ArrayInspector
  def self.inspect(array)
    array.map do |element|
      if Array === element
        ArrayInspector.inspect(element)
      else
        ObjectInspector.inspect(element)
      end
    end
  end
end
```

The only issue I see here is that a Hash inside of an Array, will have both `.inspect` called on the hash, and then again once it's in the array, which I think will have weird formatting issues.